File: Problems/0300_0399/0341_Flatten_Nested_List_Iterator/Project_Java/NestedIterator.java
import java.util.*;

// 2ms - 7ms
public class NestedIterator implements Iterator<Integer> {
    private List<Integer> integerList = new ArrayList<>();
    private int index = 0;

    public NestedIterator(List<NestedInteger> nestedList) {
        for (NestedInteger nestedInteger : nestedList) {
            flatten(nestedInteger);
        }
    }

    private void flatten(NestedInteger nested) {
        if (nested.isInteger()) {
            integerList.add(nested.getInteger());
        } else {
            for (NestedInteger nestedFromList : nested.getList()) {
                flatten(nestedFromList);
            }
        }
    }

    @Override
    public Integer next() {
        return integerList.get(index++);
    }

    @Override
    public boolean hasNext() {
        return index < integerList.size();
    }
}

/**
 * Your NestedIterator object will be instantiated and called as such:
 * NestedIterator i = new NestedIterator(nestedList);
 * while (i.hasNext()) v[f()] = i.next();
 */

File: Problems/0300_0399/0341_Flatten_Nested_List_Iterator/Project_Go/solution/nestediterator.go
package solution

// 5ms - 10ms

type NestedIterator struct {
	intList []int
	index   int
}

func Constructor(nestedList []*NestedInteger) *NestedIterator {
	ni := NestedIterator{}
	for _, nestedInteger := range nestedList {
		flatten(&ni, nestedInteger)
	}
	return &ni
}

func flatten(ni *NestedIterator, nested *NestedInteger) {
	if (*nested).IsInteger() {
		(*ni).intList = append((*ni).intList, nested.GetInteger())
	} else {
		for _, nestedFromList := range (*nested).GetList() {
			flatten(ni, nestedFromList)
		}
	}
}
func (myname *NestedIterator) Next() int {
	(*myname).index++
	return (*myname).intList[(*myname).index-1]
}

func (this *NestedIterator) HasNext() bool {
	return (*this).index < len((*this).intList)
}
